File: server/src/main/com/broll/gainea/server/core/objects/buffs/GlobalBuff.java
package com.broll.gainea.server.core.objects.buffs;

import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.player.Player;
import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class GlobalBuff {

    private Buff buff;
    private Consumer<Unit> applier;
    private List<Player> targets;

    private GlobalBuff(List<Player> targets, Buff buff, Consumer<Unit> applier) {
        this.buff = buff;
        this.applier = applier;
        this.targets = targets;
    }

    public static void createForPlayer(GameContainer game, Player target, Buff buff, Consumer<Unit> applier, int effect) {
        register(game, new GlobalBuff(Lists.newArrayList(target), buff, applier), effect);
    }

    public static void createForAllPlayers(GameContainer game, Buff buff, Consumer<Unit> applier, int effect) {
        register(game, new GlobalBuff(new ArrayList<>(game.getAllPlayers()), buff, applier), effect);
    }

    public static void createForPlayers(GameContainer game, List<Player> targets, Buff buff, Consumer<Unit> applier, int effect) {
        register(game, new GlobalBuff(new ArrayList<>(targets), buff, applier), effect);
    }

    public static void createForNeutral(GameContainer game, Buff buff, Consumer<Unit> applier, int effect) {
        register(game, new GlobalBuff(Lists.newArrayList((Player) null), buff, applier), effect);
    }

    public static void createForAll(GameContainer game, Buff buff, Consumer<Unit> applier, int effect) {
        List<Player> targets = new ArrayList<>();
        targets.addAll(game.getAllPlayers());
        targets.add(null);
        register(game, new GlobalBuff(targets, buff, applier), effect);
    }

    private static void register(GameContainer game, GlobalBuff buff, int effect) {
        game.getBuffProcessor().addGlobalBuff(buff, effect);
    }

    public Buff getBuff() {
        return buff;
    }

    public List<Player> getTargets() {
        return targets;
    }

    public void apply(Unit object) {
        applier.accept(object);
    }
}

File: server/src/main/com/broll/gainea/server/core/objects/buffs/GlobalBuff.kt
package com.broll.gainea.server.core.objects.buffs

import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.objects.Unit
import com.broll.gainea.server.core.player.Player
import java.util.function.Consumer

class GlobalBuff(
        val targets: List<Player>,
        val buff: Buff<*>,
        private val applier: Consumer<Unit>
) {

    fun apply(`object`: Unit) {
        applier.accept(`object`)
    }

    companion object {
        fun createForPlayer(game: Game, target: Player, buff: Buff<*>, applier: Consumer<Unit>, effect: Int) {
            register(game, GlobalBuff(listOf(target), buff, applier), effect)
        }

        fun createForAllPlayers(game: Game, buff: Buff<*>, applier: Consumer<Unit>, effect: Int) {
            register(game, GlobalBuff(ArrayList(game.allPlayers), buff, applier), effect)
        }

        fun createForPlayers(game: Game, targets: List<Player>, buff: Buff<*>, applier: Consumer<Unit>, effect: Int) {
            register(game, GlobalBuff(targets, buff, applier), effect)
        }

        fun createForNeutral(game: Game, buff: Buff<*>, applier: Consumer<Unit>, effect: Int) {
            register(game, GlobalBuff(listOf(game.neutralPlayer), buff, applier), effect)
        }

        fun createForAll(game: Game, buff: Buff<*>, applier: Consumer<Unit>, effect: Int) {
            register(game, GlobalBuff(listOf(game.allPlayers, listOf(game.neutralPlayer)).flatten(), buff, applier), effect)
        }

        private fun register(game: Game, buff: GlobalBuff, effect: Int) {
            game.buffProcessor.addGlobalBuff(buff, effect)
        }
    }
}
